File: apps/server/internal/compiler/sql_parser.go
package compiler

import (
	"regexp"
	"strings"
)

type SQLTable struct {
	Name    string
	Columns []SQLColumn
}

type SQLColumn struct {
	Name         string
	Type         string
	IsPrimaryKey bool
	IsUnique     bool
	IsNullable   bool
	IsForeignKey bool
	RefTable     string
	RefColumn    string
	Constraints  []string
	DefaultValue string
}

type SQLForeignKey struct {
	FromTable  string
	FromColumn string
	ToTable    string
	ToColumn   string
	Name       string
}

// ParseSQL parses SQL CREATE TABLE statements and extracts schema information
func ParseSQL(sqlContent string) ([]SQLTable, []SQLForeignKey, error) {
	tables := []SQLTable{}
	foreignKeys := []SQLForeignKey{}

	// Normalize SQL - remove comments and extra whitespace
	sqlContent = removeSQLComments(sqlContent)
	sqlContent = strings.ReplaceAll(sqlContent, "\r\n", "\n")
	sqlContent = strings.ReplaceAll(sqlContent, "\r", "\n")
	// Normalize multiple spaces
	sqlContent = regexp.MustCompile(`\s+`).ReplaceAllString(sqlContent, " ")

	// Split by semicolons to get individual statements
	statements := splitSQLStatements(sqlContent)

	for _, stmt := range statements {
		stmt = strings.TrimSpace(stmt)
		upperStmt := strings.ToUpper(stmt)

		if strings.HasPrefix(upperStmt, "CREATE TABLE") {
			table, fks := parseCreateTable(stmt)
			if table.Name != "" {
				tables = append(tables, table)
				foreignKeys = append(foreignKeys, fks...)
			}
		} else if strings.HasPrefix(upperStmt, "ALTER TABLE") {
			// Parse ALTER TABLE ADD CONSTRAINT for foreign keys
			fk := parseAlterTableFK(stmt)
			if fk != nil {
				foreignKeys = append(foreignKeys, *fk)
			}
		}
	}

	return tables, foreignKeys, nil
}

func removeSQLComments(sql string) string {
	// Remove single-line comments (-- comment)
	re := regexp.MustCompile(`--[^\n]*`)
	sql = re.ReplaceAllString(sql, "")

	// Remove multi-line comments (/* comment */)
	re = regexp.MustCompile(`(?s)/\*.*?\*/`)
	sql = re.ReplaceAllString(sql, "")

	return sql
}

func splitSQLStatements(sql string) []string {
	// Split by semicolon, but be careful with semicolons inside strings
	statements := []string{}
	current := strings.Builder{}
	inString := false
	quoteChar := byte(0)

	for i := 0; i < len(sql); i++ {
		char := sql[i]

		if (char == '\'' || char == '"' || char == '`') && (i == 0 || sql[i-1] != '\\') {
			if !inString {
				inString = true
				quoteChar = char
			} else if char == quoteChar {
				inString = false
				quoteChar = 0
			}
		}

		if char == ';' && !inString {
			stmt := strings.TrimSpace(current.String())
			if stmt != "" {
				statements = append(statements, stmt)
			}
			current.Reset()
		} else {
			current.WriteByte(char)
		}
	}

	if strings.TrimSpace(current.String()) != "" {
		statements = append(statements, strings.TrimSpace(current.String()))
	}

	return statements
}

func parseCreateTable(stmt string) (SQLTable, []SQLForeignKey) {
	table := SQLTable{}
	foreignKeys := []SQLForeignKey{}

	// Extract table name - handle various formats
	// CREATE TABLE [IF NOT EXISTS] [schema.]table_name
	re := regexp.MustCompile(`(?i)CREATE\s+TABLE\s+(?:IF\s+NOT\s+EXISTS\s+)?(?:["'\x60]?(\w+)["'\x60]?\.)?["'\x60]?(\w+)["'\x60]?\s*\(`)
	matches := re.FindStringSubmatch(stmt)
	if len(matches) < 3 {
		return table, foreignKeys
	}
	// Use schema.table or just table
	if matches[1] != "" {
		table.Name = matches[2] // Just use table name, ignore schema
	} else {
		table.Name = matches[2]
	}

	// Extract content between parentheses
	startIdx := strings.Index(stmt, "(")
	if startIdx == -1 {
		return table, foreignKeys
	}

	// Find matching closing parenthesis
	parenCount := 0
	endIdx := -1
	for i := startIdx; i < len(stmt); i++ {
		if stmt[i] == '(' {
			parenCount++
		} else if stmt[i] == ')' {
			parenCount--
			if parenCount == 0 {
				endIdx = i
				break
			}
		}
	}

	if endIdx == -1 {
		return table, foreignKeys
	}

	content := stmt[startIdx+1 : endIdx]

	// Parse column definitions and constraints
	definitions := splitColumnDefinitions(content)
	primaryKeys := []string{}
	uniqueKeys := map[string]bool{}

	for _, def := range definitions {
		def = strings.TrimSpace(def)
		if def == "" {
			continue
		}

		upperDef := strings.ToUpper(def)

		// Check for PRIMARY KEY constraint (table-level)
		if strings.HasPrefix(upperDef, "PRIMARY KEY") {
			pkRe := regexp.MustCompile(`(?i)PRIMARY\s+KEY\s*\(([^)]+)\)`)
			pkMatches := pkRe.FindStringSubmatch(def)
			if len(pkMatches) >= 2 {
				cols := strings.Split(pkMatches[1], ",")
				for _, col := range cols {
					col = cleanIdentifier(col)
					if col != "" {
						primaryKeys = append(primaryKeys, col)
					}
				}
			}
			continue
		}

		// Check for UNIQUE constraint (table-level)
		if strings.HasPrefix(upperDef, "UNIQUE") {
			uniqueRe := regexp.MustCompile(`(?i)UNIQUE\s*(?:KEY|INDEX)?\s*(?:\w+\s*)?\(([^)]+)\)`)
			uniqueMatches := uniqueRe.FindStringSubmatch(def)
			if len(uniqueMatches) >= 2 {
				cols := strings.Split(uniqueMatches[1], ",")
				for _, col := range cols {
					col = cleanIdentifier(col)
					if col != "" {
						uniqueKeys[col] = true
					}
				}
			}
			continue
		}

		// Check for FOREIGN KEY constraint (table-level)
		if strings.HasPrefix(upperDef, "FOREIGN KEY") || strings.HasPrefix(upperDef, "CONSTRAINT") {
			fk := parseForeignKeyConstraint(def, table.Name)
			if fk != nil {
				foreignKeys = append(foreignKeys, *fk)
			}
			continue
		}

		// Check for CHECK constraint
		if strings.HasPrefix(upperDef, "CHECK") {
			continue
		}

		// Check for INDEX
		if strings.HasPrefix(upperDef, "INDEX") || strings.HasPrefix(upperDef, "KEY") {
			continue
		}

		// Parse column definition
		col, inlineFk := parseColumnDefinition(def, table.Name)
		if col.Name != "" {
			table.Columns = append(table.Columns, col)
			if inlineFk != nil {
				foreignKeys = append(foreignKeys, *inlineFk)
			}
		}
	}

	// Mark primary key columns
	for i := range table.Columns {
		for _, pk := range primaryKeys {
			if strings.EqualFold(table.Columns[i].Name, pk) {
				table.Columns[i].IsPrimaryKey = true
				break
			}
		}
		// Mark unique columns
		if uniqueKeys[strings.ToLower(table.Columns[i].Name)] {
			table.Columns[i].IsUnique = true
			if !contains(table.Columns[i].Constraints, "UNQ") {
				table.Columns[i].Constraints = append(table.Columns[i].Constraints, "UNQ")
			}
		}
	}

	return table, foreignKeys
}

func splitColumnDefinitions(content string) []string {
	// Split by commas, but respect parentheses (for types like DECIMAL(10,2))
	definitions := []string{}
	current := strings.Builder{}
	parenCount := 0
	inString := false
	quoteChar := byte(0)

	for i := 0; i < len(content); i++ {
		char := content[i]

		// Handle string literals
		if (char == '\'' || char == '"' || char == '`') && (i == 0 || content[i-1] != '\\') {
			if !inString {
				inString = true
				quoteChar = char
			} else if char == quoteChar {
				inString = false
				quoteChar = 0
			}
		}

		if !inString {
			if char == '(' {
				parenCount++
			} else if char == ')' {
				parenCount--
			}
		}

		if char == ',' && parenCount == 0 && !inString {
			def := strings.TrimSpace(current.String())
			if def != "" {
				definitions = append(definitions, def)
			}
			current.Reset()
		} else {
			current.WriteByte(char)
		}
	}

	if strings.TrimSpace(current.String()) != "" {
		definitions = append(definitions, strings.TrimSpace(current.String()))
	}

	return definitions
}

func parseColumnDefinition(def string, tableName string) (SQLColumn, *SQLForeignKey) {
	col := SQLColumn{
		IsNullable:  true,
		Constraints: []string{},
	}
	var fk *SQLForeignKey

	// Tokenize the definition
	tokens := tokenizeColumnDef(def)
	if len(tokens) == 0 {
		return col, nil
	}

	// First token is column name
	col.Name = cleanIdentifier(tokens[0])
	if col.Name == "" {
		return col, nil
	}

	// Find data type (next token after name)
	if len(tokens) > 1 {
		col.Type = normalizeDataType(tokens[1:])
	}

	// Parse constraints from the rest of the definition
	upperDef := strings.ToUpper(def)

	// Check for NOT NULL
	if strings.Contains(upperDef, "NOT NULL") {
		col.IsNullable = false
		col.Constraints = append(col.Constraints, "NN")
	}

	// Check for UNIQUE
	if regexp.MustCompile(`(?i)\bUNIQUE\b`).MatchString(def) {
		col.IsUnique = true
		if !contains(col.Constraints, "UNQ") {
			col.Constraints = append(col.Constraints, "UNQ")
		}
	}

	// Check for PRIMARY KEY
	if regexp.MustCompile(`(?i)\bPRIMARY\s+KEY\b`).MatchString(def) {
		col.IsPrimaryKey = true
	}

	// Check for AUTO_INCREMENT / SERIAL / IDENTITY
	if regexp.MustCompile(`(?i)\b(AUTO_INCREMENT|SERIAL|BIGSERIAL|SMALLSERIAL|IDENTITY)\b`).MatchString(def) {
		if !contains(col.Constraints, "AI") {
			col.Constraints = append(col.Constraints, "AI")
		}
	}

	// Check for DEFAULT value
	defaultRe := regexp.MustCompile(`(?i)\bDEFAULT\s+([^\s,]+(?:\([^)]*\))?)`)
	defaultMatches := defaultRe.FindStringSubmatch(def)
	if len(defaultMatches) >= 2 {
		col.DefaultValue = defaultMatches[1]
	}

	// Check for inline REFERENCES (foreign key)
	refRe := regexp.MustCompile(`(?i)\bREFERENCES\s+["'\x60]?(\w+)["'\x60]?\s*\(\s*["'\x60]?(\w+)["'\x60]?\s*\)`)
	refMatches := refRe.FindStringSubmatch(def)
	if len(refMatches) >= 3 {
		col.IsForeignKey = true
		col.RefTable = refMatches[1]
		col.RefColumn = refMatches[2]
		if !contains(col.Constraints, "FK") {
			col.Constraints = append(col.Constraints, "FK")
		}
		fk = &SQLForeignKey{
			FromTable:  tableName,
			FromColumn: col.Name,
			ToTable:    refMatches[1],
			ToColumn:   refMatches[2],
		}
	}

	return col, fk
}

func tokenizeColumnDef(def string) []string {
	tokens := []string{}
	current := strings.Builder{}
	inString := false
	quoteChar := byte(0)
	parenCount := 0

	for i := 0; i < len(def); i++ {
		char := def[i]

		// Handle string literals
		if (char == '\'' || char == '"' || char == '`') && (i == 0 || def[i-1] != '\\') {
			if !inString {
				inString = true
				quoteChar = char
			} else if char == quoteChar {
				inString = false
				quoteChar = 0
			}
			current.WriteByte(char)
			continue
		}

		if !inString {
			if char == '(' {
				parenCount++
				current.WriteByte(char)
			} else if char == ')' {
				parenCount--
				current.WriteByte(char)
			} else if char == ' ' || char == '\t' || char == '\n' {
				if parenCount > 0 {
					// Inside parentheses, keep spaces for types like DECIMAL(10, 2)
					current.WriteByte(char)
				} else if current.Len() > 0 {
					tokens = append(tokens, current.String())
					current.Reset()
				}
			} else {
				current.WriteByte(char)
			}
		} else {
			current.WriteByte(char)
		}
	}

	if current.Len() > 0 {
		tokens = append(tokens, current.String())
	}

	return tokens
}

func normalizeDataType(tokens []string) string {
	if len(tokens) == 0 {
		return "text"
	}

	typeStr := strings.ToUpper(tokens[0])

	// Handle types with parameters
	if strings.Contains(typeStr, "(") {
		return normalizeTypeString(typeStr)
	}

	// Check if next token is a parameter
	if len(tokens) > 1 && strings.HasPrefix(tokens[1], "(") {
		return normalizeTypeString(typeStr + tokens[1])
	}

	// Map common types
	return normalizeTypeString(typeStr)
}

func normalizeTypeString(typeStr string) string {
	typeStr = strings.ToUpper(typeStr)

	// PostgreSQL types
	switch {
	case strings.HasPrefix(typeStr, "SERIAL"):
		return "integer"
	case strings.HasPrefix(typeStr, "BIGSERIAL"):
		return "bigint"
	case strings.HasPrefix(typeStr, "SMALLSERIAL"):
		return "integer"
	case typeStr == "UUID":
		return "uuid"
	case strings.HasPrefix(typeStr, "VARCHAR"):
		return strings.ToLower(typeStr)
	case strings.HasPrefix(typeStr, "CHARACTER VARYING"):
		return "varchar(255)"
	case typeStr == "TEXT":
		return "text"
	case typeStr == "INTEGER" || typeStr == "INT":
		return "integer"
	case typeStr == "BIGINT":
		return "bigint"
	case typeStr == "SMALLINT":
		return "integer"
	case typeStr == "BOOLEAN" || typeStr == "BOOL":
		return "boolean"
	case strings.HasPrefix(typeStr, "TIMESTAMP"):
		return "timestamp"
	case typeStr == "DATE":
		return "date"
	case typeStr == "TIME":
		return "timestamp"
	case strings.HasPrefix(typeStr, "NUMERIC") || strings.HasPrefix(typeStr, "DECIMAL"):
		if strings.Contains(typeStr, "(") {
			return strings.ToLower(typeStr)
		}
		return "decimal(10,2)"
	case typeStr == "REAL" || typeStr == "FLOAT4":
		return "decimal(10,2)"
	case typeStr == "DOUBLE PRECISION" || typeStr == "FLOAT8":
		return "decimal(10,2)"
	case typeStr == "JSONB" || typeStr == "JSON":
		return "jsonb"
	case typeStr == "BYTEA" || typeStr == "BLOB":
		return "text"
	case strings.HasPrefix(typeStr, "CHAR("):
		return strings.ToLower(typeStr)
	case typeStr == "CHAR":
		return "varchar(255)"
	// MySQL types
	case strings.HasPrefix(typeStr, "INT("):
		return "integer"
	case typeStr == "TINYINT":
		return "integer"
	case typeStr == "MEDIUMINT":
		return "integer"
	case strings.HasPrefix(typeStr, "TINYINT"):
		return "boolean"
	case typeStr == "DATETIME":
		return "timestamp"
	case strings.HasPrefix(typeStr, "ENUM"):
		return "varchar(255)"
	case strings.HasPrefix(typeStr, "SET"):
		return "varchar(255)"
	default:
		return strings.ToLower(typeStr)
	}
}

func parseForeignKeyConstraint(def string, tableName string) *SQLForeignKey {
	// Pattern: [CONSTRAINT name] FOREIGN KEY (column[, ...]) REFERENCES ref_table(ref_column[, ...])
	re := regexp.MustCompile(`(?i)(?:CONSTRAINT\s+["'\x60]?(\w+)["'\x60]?\s+)?FOREIGN\s+KEY\s*\(\s*["'\x60]?(\w+)["'\x60]?\s*\)\s+REFERENCES\s+["'\x60]?(\w+)["'\x60]?\s*\(\s*["'\x60]?(\w+)["'\x60]?\s*\)`)
	matches := re.FindStringSubmatch(def)
	if len(matches) >= 5 {
		return &SQLForeignKey{
			Name:       matches[1],
			FromTable:  tableName,
			FromColumn: matches[2],
			ToTable:    matches[3],
			ToColumn:   matches[4],
		}
	}
	return nil
}

func parseAlterTableFK(stmt string) *SQLForeignKey {
	// Pattern: ALTER TABLE table_name ADD [CONSTRAINT constraint_name] FOREIGN KEY (column) REFERENCES ref_table(ref_column)
	re := regexp.MustCompile(`(?i)ALTER\s+TABLE\s+["'\x60]?(\w+)["'\x60]?\s+ADD\s+(?:CONSTRAINT\s+["'\x60]?\w+["'\x60]?\s+)?FOREIGN\s+KEY\s*\(\s*["'\x60]?(\w+)["'\x60]?\s*\)\s+REFERENCES\s+["'\x60]?(\w+)["'\x60]?\s*\(\s*["'\x60]?(\w+)["'\x60]?\s*\)`)
	matches := re.FindStringSubmatch(stmt)
	if len(matches) >= 5 {
		return &SQLForeignKey{
			FromTable:  matches[1],
			FromColumn: matches[2],
			ToTable:    matches[3],
			ToColumn:   matches[4],
		}
	}
	return nil
}

func cleanIdentifier(s string) string {
	s = strings.TrimSpace(s)
	s = strings.Trim(s, "`\"'")
	return s
}

func contains(slice []string, item string) bool {
	for _, s := range slice {
		if s == item {
			return true
		}
	}
	return false
}

File: apps/web/lib/compiler/sql-parser.ts
// Port of internal/compiler/sql_parser.go — pattern-for-pattern.

export interface SQLColumn {
  name: string;
  type: string;
  isPrimaryKey: boolean;
  isUnique: boolean;
  isNullable: boolean;
  isForeignKey: boolean;
  refTable?: string;
  refColumn?: string;
  constraints: string[];
  defaultValue?: string;
}

export interface SQLTable {
  name: string;
  columns: SQLColumn[];
}

export interface SQLForeignKey {
  fromTable: string;
  fromColumn: string;
  toTable: string;
  toColumn: string;
  name?: string;
}

function removeSQLComments(sql: string): string {
  return sql.replace(/--[^\n]*/g, "").replace(/\/\*[\s\S]*?\*\//g, "");
}

function splitSQLStatements(sql: string): string[] {
  const out: string[] = [];
  let current = "";
  let inString = false;
  let quote = "";
  for (let i = 0; i < sql.length; i++) {
    const ch = sql[i]!;
    if ((ch === "'" || ch === '"' || ch === "`") && (i === 0 || sql[i - 1] !== "\\")) {
      if (!inString) {
        inString = true;
        quote = ch;
      } else if (ch === quote) {
        inString = false;
        quote = "";
      }
    }
    if (ch === ";" && !inString) {
      const t = current.trim();
      if (t) out.push(t);
      current = "";
    } else {
      current += ch;
    }
  }
  const last = current.trim();
  if (last) out.push(last);
  return out;
}

function cleanIdentifier(s: string): string {
  return s.trim().replace(/^["'`]+|["'`]+$/g, "");
}

function splitColumnDefinitions(content: string): string[] {
  const out: string[] = [];
  let current = "";
  let parens = 0;
  let inString = false;
  let quote = "";
  for (let i = 0; i < content.length; i++) {
    const ch = content[i]!;
    if ((ch === "'" || ch === '"' || ch === "`") && (i === 0 || content[i - 1] !== "\\")) {
      if (!inString) {
        inString = true;
        quote = ch;
      } else if (ch === quote) {
        inString = false;
        quote = "";
      }
    }
    if (!inString) {
      if (ch === "(") parens++;
      else if (ch === ")") parens--;
    }
    if (ch === "," && parens === 0 && !inString) {
      const t = current.trim();
      if (t) out.push(t);
      current = "";
    } else {
      current += ch;
    }
  }
  const last = current.trim();
  if (last) out.push(last);
  return out;
}

function tokenizeColumnDef(def: string): string[] {
  const tokens: string[] = [];
  let current = "";
  let inString = false;
  let quote = "";
  let parens = 0;
  for (let i = 0; i < def.length; i++) {
    const ch = def[i]!;
    if ((ch === "'" || ch === '"' || ch === "`") && (i === 0 || def[i - 1] !== "\\")) {
      if (!inString) {
        inString = true;
        quote = ch;
      } else if (ch === quote) {
        inString = false;
        quote = "";
      }
      current += ch;
      continue;
    }
    if (!inString) {
      if (ch === "(") {
        parens++;
        current += ch;
      } else if (ch === ")") {
        parens--;
        current += ch;
      } else if (ch === " " || ch === "\t" || ch === "\n") {
        if (parens > 0) current += ch;
        else if (current.length) {
          tokens.push(current);
          current = "";
        }
      } else current += ch;
    } else current += ch;
  }
  if (current.length) tokens.push(current);
  return tokens;
}

function normalizeTypeString(typeStr: string): string {
  const u = typeStr.toUpperCase();
  if (u.startsWith("SERIAL")) return "integer";
  if (u.startsWith("BIGSERIAL")) return "bigint";
  if (u.startsWith("SMALLSERIAL")) return "integer";
  if (u === "UUID") return "uuid";
  if (u.startsWith("VARCHAR")) return u.toLowerCase();
  if (u.startsWith("CHARACTER VARYING")) return "varchar(255)";
  if (u === "TEXT") return "text";
  if (u === "INTEGER" || u === "INT") return "integer";
  if (u === "BIGINT") return "bigint";
  if (u === "SMALLINT") return "integer";
  if (u === "BOOLEAN" || u === "BOOL") return "boolean";
  if (u.startsWith("TIMESTAMP")) return "timestamp";
  if (u === "DATE") return "date";
  if (u === "TIME") return "timestamp";
  if (u.startsWith("NUMERIC") || u.startsWith("DECIMAL")) {
    return u.includes("(") ? u.toLowerCase() : "decimal(10,2)";
  }
  if (u === "REAL" || u === "FLOAT4") return "decimal(10,2)";
  if (u === "DOUBLE PRECISION" || u === "FLOAT8") return "decimal(10,2)";
  if (u === "JSONB" || u === "JSON") return "jsonb";
  if (u === "BYTEA" || u === "BLOB") return "text";
  if (u.startsWith("CHAR(")) return u.toLowerCase();
  if (u === "CHAR") return "varchar(255)";
  if (u.startsWith("INT(")) return "integer";
  if (u === "TINYINT" || u === "MEDIUMINT") return "integer";
  if (u.startsWith("TINYINT")) return "boolean";
  if (u === "DATETIME") return "timestamp";
  if (u.startsWith("ENUM") || u.startsWith("SET")) return "varchar(255)";
  return u.toLowerCase();
}

function normalizeDataType(tokens: string[]): string {
  if (!tokens.length) return "text";
  let typeStr = tokens[0]!.toUpperCase();
  if (typeStr.includes("(")) return normalizeTypeString(typeStr);
  if (tokens.length > 1 && tokens[1]!.startsWith("(")) {
    return normalizeTypeString(typeStr + tokens[1]);
  }
  return normalizeTypeString(typeStr);
}

function parseColumnDefinition(
  def: string,
  tableName: string,
): { col: SQLColumn; fk?: SQLForeignKey } {
  const col: SQLColumn = {
    name: "",
    type: "",
    isPrimaryKey: false,
    isUnique: false,
    isNullable: true,
    isForeignKey: false,
    constraints: [],
  };
  const tokens = tokenizeColumnDef(def);
  if (!tokens.length) return { col };

  col.name = cleanIdentifier(tokens[0]!);
  if (!col.name) return { col };

  if (tokens.length > 1) col.type = normalizeDataType(tokens.slice(1));

  const upper = def.toUpperCase();
  if (upper.includes("NOT NULL")) {
    col.isNullable = false;
    col.constraints.push("NN");
  }
  if (/\bUNIQUE\b/i.test(def)) {
    col.isUnique = true;
    if (!col.constraints.includes("UNQ")) col.constraints.push("UNQ");
  }
  if (/\bPRIMARY\s+KEY\b/i.test(def)) col.isPrimaryKey = true;
  if (/\b(AUTO_INCREMENT|SERIAL|BIGSERIAL|SMALLSERIAL|IDENTITY)\b/i.test(def)) {
    if (!col.constraints.includes("AI")) col.constraints.push("AI");
  }
  const defaultMatch = def.match(/\bDEFAULT\s+([^\s,]+(?:\([^)]*\))?)/i);
  if (defaultMatch) col.defaultValue = defaultMatch[1];

  const refMatch = def.match(/\bREFERENCES\s+["'`]?(\w+)["'`]?\s*\(\s*["'`]?(\w+)["'`]?\s*\)/i);
  let fk: SQLForeignKey | undefined;
  if (refMatch) {
    col.isForeignKey = true;
    col.refTable = refMatch[1];
    col.refColumn = refMatch[2];
    if (!col.constraints.includes("FK")) col.constraints.push("FK");
    fk = {
      fromTable: tableName,
      fromColumn: col.name,
      toTable: refMatch[1]!,
      toColumn: refMatch[2]!,
    };
  }

  return { col, fk };
}

function parseForeignKeyConstraint(def: string, tableName: string): SQLForeignKey | undefined {
  const re =
    /(?:CONSTRAINT\s+["'`]?(\w+)["'`]?\s+)?FOREIGN\s+KEY\s*\(\s*["'`]?(\w+)["'`]?\s*\)\s+REFERENCES\s+["'`]?(\w+)["'`]?\s*\(\s*["'`]?(\w+)["'`]?\s*\)/i;
  const m = def.match(re);
  if (!m) return undefined;
  return {
    name: m[1],
    fromTable: tableName,
    fromColumn: m[2]!,
    toTable: m[3]!,
    toColumn: m[4]!,
  };
}

function parseAlterTableFK(stmt: string): SQLForeignKey | undefined {
  const re =
    /ALTER\s+TABLE\s+["'`]?(\w+)["'`]?\s+ADD\s+(?:CONSTRAINT\s+["'`]?\w+["'`]?\s+)?FOREIGN\s+KEY\s*\(\s*["'`]?(\w+)["'`]?\s*\)\s+REFERENCES\s+["'`]?(\w+)["'`]?\s*\(\s*["'`]?(\w+)["'`]?\s*\)/i;
  const m = stmt.match(re);
  if (!m) return undefined;
  return {
    fromTable: m[1]!,
    fromColumn: m[2]!,
    toTable: m[3]!,
    toColumn: m[4]!,
  };
}

function parseCreateTable(stmt: string): { table: SQLTable; fks: SQLForeignKey[] } {
  const fks: SQLForeignKey[] = [];
  const table: SQLTable = { name: "", columns: [] };

  const headRe =
    /CREATE\s+TABLE\s+(?:IF\s+NOT\s+EXISTS\s+)?(?:["'`]?(\w+)["'`]?\.)?["'`]?(\w+)["'`]?\s*\(/i;
  const m = stmt.match(headRe);
  if (!m) return { table, fks };
  table.name = m[2]!;

  const startIdx = stmt.indexOf("(");
  if (startIdx === -1) return { table, fks };
  let depth = 0;
  let endIdx = -1;
  for (let i = startIdx; i < stmt.length; i++) {
    if (stmt[i] === "(") depth++;
    else if (stmt[i] === ")") {
      depth--;
      if (depth === 0) {
        endIdx = i;
        break;
      }
    }
  }
  if (endIdx === -1) return { table, fks };

  const content = stmt.slice(startIdx + 1, endIdx);
  const defs = splitColumnDefinitions(content);
  const primaryKeys: string[] = [];
  const uniqueKeys = new Set<string>();

  for (const raw of defs) {
    const def = raw.trim();
    if (!def) continue;
    const upper = def.toUpperCase();

    if (upper.startsWith("PRIMARY KEY")) {
      const pk = def.match(/PRIMARY\s+KEY\s*\(([^)]+)\)/i);
      if (pk) for (const c of pk[1]!.split(",")) {
        const id = cleanIdentifier(c);
        if (id) primaryKeys.push(id);
      }
      continue;
    }
    if (upper.startsWith("UNIQUE")) {
      const u = def.match(/UNIQUE\s*(?:KEY|INDEX)?\s*(?:\w+\s*)?\(([^)]+)\)/i);
      if (u) for (const c of u[1]!.split(",")) {
        const id = cleanIdentifier(c);
        if (id) uniqueKeys.add(id.toLowerCase());
      }
      continue;
    }
    if (upper.startsWith("FOREIGN KEY") || upper.startsWith("CONSTRAINT")) {
      const fk = parseForeignKeyConstraint(def, table.name);
      if (fk) fks.push(fk);
      continue;
    }
    if (upper.startsWith("CHECK") || upper.startsWith("INDEX") || upper.startsWith("KEY")) continue;

    const { col, fk } = parseColumnDefinition(def, table.name);
    if (col.name) {
      table.columns.push(col);
      if (fk) fks.push(fk);
    }
  }

  for (const c of table.columns) {
    if (primaryKeys.some((pk) => pk.toLowerCase() === c.name.toLowerCase())) c.isPrimaryKey = true;
    if (uniqueKeys.has(c.name.toLowerCase())) {
      c.isUnique = true;
      if (!c.constraints.includes("UNQ")) c.constraints.push("UNQ");
    }
  }

  return { table, fks };
}

export function parseSQL(sqlContent: string): { tables: SQLTable[]; foreignKeys: SQLForeignKey[] } {
  let sql = removeSQLComments(sqlContent);
  sql = sql.replace(/\r\n/g, "\n").replace(/\r/g, "\n").replace(/\s+/g, " ");
  const stmts = splitSQLStatements(sql);

  const tables: SQLTable[] = [];
  const foreignKeys: SQLForeignKey[] = [];

  for (const raw of stmts) {
    const stmt = raw.trim();
    const upper = stmt.toUpperCase();
    if (upper.startsWith("CREATE TABLE")) {
      const { table, fks } = parseCreateTable(stmt);
      if (table.name) {
        tables.push(table);
        foreignKeys.push(...fks);
      }
    } else if (upper.startsWith("ALTER TABLE")) {
      const fk = parseAlterTableFK(stmt);
      if (fk) foreignKeys.push(fk);
    }
  }

  return { tables, foreignKeys };
}
